rfc/004-priority-based-mempool.md

```diff
@@ -0,0 +1,205 @@
+
+The typical flow is:
+
+![mempool](https://user-images.githubusercontent.com/1282182/92575349-42a13280-f299-11ea-8b6e-bd7f27d536e6.png)
+
+On the other hand, user-defined priorities for blockchain transactions can be useful, e.g. `gasPrice` in Ethereum blockchain works as a first-price auction mechanism to cope with limited capacity. However in Tendermint, current approach for packing transactions into a block follows a first-in-first-out (FIFO) style which is baked in Tendermint engine and can't be affected by ABCI applications.
+
+To make Tendermint support priority-based transaction processing, two main aspects of current architecture in Tendermint need to be updated: one is on ABCI's interface for accepting a transaction (`CheckTx`), and another is on Tendermint's mempool design and implementation. This RFC proposes corresponding changes to address these two aspects.
+
+## Current implementation and its flaws
+
+In the heart of the current implementation there are a concurrent linked list and a map for a quick access (key - SHA256 hash of the transaction). The list is unordered. The map was added at a much later stage when we needed to record/retrieve the sender (in order to avoid sending the transaction back, which was wasteful).
+
+The mempool also contains the LRU cache of the most recently used transactions. The cache prevents same transactions from entering the mempool and serves as a basis layer of replay protection for the ABCI application. However, the ABCI applications must implement the additional means of protection in case the cache is not enough.
+
+![mempool2](https://user-images.githubusercontent.com/1282182/92575388-52b91200-f299-11ea-91c8-10e05e7be4df.png)
+
+It’s been reported that the linked list and the map could [diverge after the cache becomes full](https://github.com/tendermint/tendermint/issues/5281).
+
+The biggest downside of the current implementation is the lack of prioritization. It should give the ABCI application a way to establish some order, based on the fee or some other field(s).
+
+Dev and others also expressed some concerns about concurrency patters - “My concern is primarily that we currently have the downside that you can't keep receiving / broadcasting txs while updating / rechecking.“ (see [this issue](https://github.com/tendermint/tendermint/issues/2147) and [this one](https://github.com/tendermint/tendermint/issues/2484))
+
+Other possible optimizations:
+
+- [highwayhash instead of SHA256](https://github.com/tendermint/tendermint/issues/2187)
+- [remove storing height](https://github.com/tendermint/tendermint/issues/2834)
+- [remove gasWanted for non-validators](https://github.com/tendermint/tendermint/issues/2978)
+- [time-based eviction](https://github.com/tendermint/tendermint/issues/3436)
+
+What we want from the new design:
```

Comment:
I think it would be helpful to have a clear more comprehensive statement of the functioning of the mempool. It should include the need to:

- Broadcast txs (in order) to peers, concurrently and efficiently (ie. without resending the same txs many times)
- Execute CheckTx serially but concurrently with other Tendermint functionality (like executing a block)
- Allow proposers to efficiently reap txs from the mempool (possibly concurrently with new CheckTxs happening?)
- Support tx rechecking and re-prioritization (?!) after a block is executed 
- Ensure priority ordered txs still form a valid ordering (ie. they pass CheckTx in that order and thus would pass DeliverTx in that order)
- 
- More?

The mempool has to handle a diverse range of concerns and it would be good to get clarity on them upfront so we can weigh the full set of constraints when evaluating designs. 

It also might give us an opportunity to realize that certain requirements are actually not necessary, or certain other requirements are necessary.

Reply:
> Ensure priority ordered txs still form a valid ordering (ie. they pass CheckTx in that order and thus would pass DeliverTx in that order)

Does this mean we'll have to re-execute CheckTx on the whole mempool every time we sort it? Otherwise, how would we guarantee that a sequence will pass?

> re-prioritization (?!)

we won't be needing this if we choose a tree, which is always sorted (actual)

Reply:
> Does this mean we'll have to re-execute CheckTx on the whole mempool every time we sort it? Otherwise, how would we guarantee that a sequence will pass?

This is going to be a challenge. We'll have to figure something out here. Possibly the mempool will have to be more aware of the relationship between txs (eg if they pertain to the same "account", whatever that might be from tendermint's perspective). One way to possibly do this is leave it to the app to return a list of txs that are invalidated by a new tx, for instance ... (?)

Reply:
> Ensure priority ordered txs still form a valid ordering (ie. they pass CheckTx in that order and thus would pass DeliverTx in that order)

where is this requirement coming from?

Reply:
From the fact that we need proposers to proposer blocks that have a sequence of validly ordered txs. Like if I send a gaia tx with sequence 5 and a small fee, and then a nother tx with sequence 6 and a large fee, the sequence 5 tx still needs to come first! (in bitcoin land they might call this "child pays for parent")

Reply:
Right. Then we might need that "account" idea when `CheckTx` optionally returns a grouping key ("account adress") and a sequence number. The proposer would need to replace a transaction with lower sequence number if such exists.